syntax = "proto3";

package streaming;

// Сервис для стриминга аудио и текста
service StreamingService {
  // Стриминг аудио и текста в ответ на промпт
  rpc StreamAudio(StreamRequest) returns (stream StreamResponse);
  
  // ПРИНУДИТЕЛЬНОЕ прерывание активной сессии на сервере
  rpc InterruptSession(InterruptRequest) returns (InterruptResponse);
  
  // Генерация приветственного аудио сообщения
  rpc GenerateWelcomeAudio(WelcomeRequest) returns (stream WelcomeResponse);

  // Репорт использования токенов (например, от клиента)
  rpc ReportUsage(UsageRequest) returns (UsageResponse);
}

// Запрос на репорт использования токенов
message UsageRequest {
  string hardware_id = 1;     // Hardware ID
  string session_id = 2;      // Session ID
  string source = 3;          // Источник (e.g. "browser_agent")
  int32 input_tokens = 4;     // Входные токены
  int32 output_tokens = 5;    // Выходные токены
  string model = 6;           // Модель
}

// Ответ на репорт
message UsageResponse {
  bool success = 1;
  string message = 2;
}

// Запрос на стриминг
message StreamRequest {
  string prompt = 1;           // Текстовая команда пользователя
  optional string screenshot = 2;       // Base64 WebP скриншот экрана (опционально)
  optional int32 screen_width = 3;     // Ширина экрана
  optional int32 screen_height = 4;    // Высота экрана
  string hardware_id = 5;      // REQUIRED: Уникальный Hardware ID оборудования (не может быть пустым или "unknown")
  string session_id = 6;      // REQUIRED: ID сессии для отслеживания (обязателен с 2026-01-12, см. CHANGELOG.md)
}

// Ответ стриминга
message StreamResponse {
  oneof content {
    string text_chunk = 1;     // Текстовый чанк
    AudioChunk audio_chunk = 2; // Аудио чанк
    string end_message = 3;    // Сообщение о завершении
    string error_message = 4;  // Сообщение об ошибке
    ActionMessage action_message = 5; // MCP/Action payload (e.g. open_app)
    BrowserProgressMessage browser_progress = 6; // browser-use automation progress
  }
}

// Аудио чанк
message AudioChunk {
  bytes audio_data = 1;        // Аудио данные
  string dtype = 2;            // Тип данных (например, 'int16')
  repeated int32 shape = 3;    // Форма массива
  int32 sample_rate = 4;       // Частота дискретизации (Hz)
  int32 channels = 5;          // Количество каналов
}

// Запрос на прерывание сессии
message InterruptRequest {
  string hardware_id = 1;      // Hardware ID для прерывания
}

// Ответ на прерывание сессии
message InterruptResponse {
  bool success = 1;            // Успешность операции
  repeated string interrupted_sessions = 2;  // Список прерванных сессий
  string message = 3;          // Сообщение о результате
}

// Запрос на генерацию приветственного аудио
message WelcomeRequest {
  string text = 1;            // Текст для генерации (опционально, используется дефолтное приветствие)
  string session_id = 2;       // ID сессии для отслеживания (опционально)
  string voice = 3;           // Голос для TTS (опционально)
  string language = 4;        // Язык для TTS (опционально)
}

// Метаданные приветственного аудио
message WelcomeMetadata {
  string method = 1;          // Метод генерации (например, "azure_tts")
  double duration_sec = 2;    // Длительность аудио в секундах
  int32 sample_rate = 3;      // Частота дискретизации
  int32 channels = 4;         // Количество каналов
  string dtype = 5;           // Тип данных (например, 'int16')
}

// Ответ на генерацию приветственного аудио
message WelcomeResponse {
  oneof content {
    AudioChunk audio_chunk = 1;  // Аудио чанк
    WelcomeMetadata metadata = 2; // Метаданные (отправляется один раз в начале)
    string end_message = 3;      // Сообщение о завершении
    string error_message = 4;    // Сообщение об ошибке
  }
}

// Action message for MCP/OpenApp
message ActionMessage {
  string action_json = 1;      // Canonical JSON: {"command":"open_app","args":{"app_name":"Safari"}}
  string session_id = 2;
  optional string feature_id = 3;
}

// Browser-use automation progress
enum BrowserEventType {
  BROWSER_TASK_STARTED = 0;
  BROWSER_STEP_STARTED = 1;
  BROWSER_STEP_COMPLETED = 2;
  BROWSER_ACTION_EXECUTED = 3;
  BROWSER_TASK_COMPLETED = 4;
  BROWSER_TASK_FAILED = 5;
  BROWSER_TASK_CANCELLED = 6;
}

message BrowserProgressMessage {
  BrowserEventType type = 1;
  string task_id = 2;
  optional int32 step_number = 3;
  optional string description = 4;  // Human-readable description
  optional string url = 5;          // Current URL
  optional string action = 6;       // Action type: go_to_url, type, click
  string timestamp = 7;             // ISO 8601 timestamp
  optional string error = 8;        // Error message if BROWSER_TASK_FAILED
  optional BrowserProgressDetails details = 9;
}

message BrowserProgressDetails {
  optional double duration_sec = 1;  // Duration in seconds
  repeated string actions = 2;       // List of actions
  map<string, string> metadata = 3;  // Additional metadata
}
